### PR TITLE
KOGITO-8028: Dashbuilder Viewer VSCode Extension

### PR DESCRIPTION
**Pull request by jesuino**

### CODE REVIEW

**Comment by jesuino** [comment]
To test:

* Download and install VSIX
* Clone [1], import into VSCode;
* Now you should have the option to open files with "Dashbuilder Editor"
* Using the command "Dashbuilder: Setup automatically open Dashboard Editor alongside Text Editor" you can configure VSCode to always open Dashbuilder editor on the side 

image

Answering "yes" will make Dashbuilder always open on the side:

image

* Bear in mind that some dashboards are broke locally, but just opening it should be fine

[1] https://github.com/jesuino/dashbuilder-yaml-samples

File: packages/vscode-extension-dashbuilder-editor/package.json [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  },
+  "devDependencies": {
+    "@kie-tools-core/webpack-base": "workspace:*",
+    "@kie-tools/dashbuilder-client": "workspace:*",
+    "@kie-tools/eslint": "workspace:*",
+    "@kie-tools/root-env": "workspace:*",
+    "@kie-tools/tsconfig": "workspace:*",
+    "@types/react": "^17.0.6",
+    "@types/vscode": "^1.66.0",
+    "@vscode/test-web": "^0.0.22",
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/package.json]: Please use version `^0.0.30` like the other packages.
```suggestion
    "@vscode/test-web": "^0.0.30",
```

File: packages/dashbuilder/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/ExternalComponentView.java [hunk 2]
```diff
@@ -66,7 +65,7 @@ public void init(ExternalComponentPresenter presenter) {
 
     @Override
     public void setComponentURL(String url) {
-        externalComponentIFrame.src = url;
+        externalComponentIFrame.src = url;
```
Comment (caponetto) [inline packages/dashbuilder/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/ExternalComponentView.java]:
```suggestion
        externalComponentIFrame.src = url;
```
Please remove these extra whitespaces. 😛

File: packages/vscode-extension-dashbuilder-editor/package.json [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    "@kie-tools-core/i18n": "workspace:*",
+    "@kie-tools-core/notifications": "workspace:*",
+    "@kie-tools-core/patternfly-base": "workspace:*",
+    "@kie-tools-core/vscode-extension": "workspace:*",
+    "@kie-tools-core/vscode-java-code-completion": "workspace:*",
+    "@kie-tools-core/workspace": "workspace:*",
+    "@kie-tools/dashbuilder-viewer": "workspace:*",
+    "portfinder": "^1.0.32",
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2"
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/package.json]: I think these react deps are not necessary.

File: pnpm-lock.yaml [hunk 4]
```diff
@@ -10597,6 +10660,28 @@ packages:
       { integrity: sha512-sL/cEvJWAnClXw0wHk85/2L0G6Sj8UB0Ctc1TEMbKSsmpRosqhwj9gWgFRZSrBr2f9tiXISwNhCPmlfqUqyb9Q== }
     dev: true
 
+  /@vscode/test-web/0.0.22:
```
Comment (caponetto) [inline pnpm-lock.yaml]: This will be removed after you update to `v0.0.30`.

File: packages/vscode-extension-dashbuilder-editor/webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+module.exports = async (env) => [
+  merge(commonConfig(env), {
+    target: "node",
+    entry: {
+      "extension/extension": "./src/extension/extension.ts",
+    },
+    plugins: [],
+  }),
+  merge(commonConfig(env), {
+    target: "web",
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/webpack.config.js]:
```suggestion
    target: "webworker",
```
I believe this needs to be `webworker`.

File: packages/vscode-extension-dashbuilder-editor/package.json [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    "@kie-tools-core/envelope-bus": "workspace:*",
+    "@kie-tools-core/guided-tour": "workspace:*",
+    "@kie-tools-core/i18n": "workspace:*",
+    "@kie-tools-core/notifications": "workspace:*",
+    "@kie-tools-core/patternfly-base": "workspace:*",
+    "@kie-tools-core/vscode-extension": "workspace:*",
+    "@kie-tools-core/vscode-java-code-completion": "workspace:*",
+    "@kie-tools-core/workspace": "workspace:*",
+    "@kie-tools/dashbuilder-viewer": "workspace:*",
+    "portfinder": "^1.0.32",
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/package.json]: Can you please update the `portfinder` version on `@kie-tools-core/backend` to `^1.0.32` as well? 
Or use here the version `^1.0.27`. Versions must match.

File: packages/vscode-extension-dashbuilder-editor/src/browser/extension.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  const kieEditorsStore = await KogitoVsCode.startExtension({
+    extensionName: "kie-group.vscode-extension-dashbuilder-editor",
+    context: context,
+    editorDocumentType: "text",
+    viewType: "kieKogitoWebviewEditorsDashbuilder",
+    editorEnvelopeLocator: new EditorEnvelopeLocator("vscode", [
+      new EnvelopeMapping({
+        type: "dashbuilder",
+        filePathGlob: "**/*.dash.+(yaml|yml)",
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/src/browser/extension.ts]:
```suggestion
        filePathGlob: "**/*.dash.+(yaml|yml|json)",
```
Can you put `json` here and remove the `EnvelopeMapping` below?

File: packages/vscode-extension-dashbuilder-editor/src/extension/extension.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+  const kieEditorsStore = await KogitoVsCode.startExtension({
+    extensionName: "kie-group.vscode-extension-dashbuilder-editor",
+    context: context,
+    editorDocumentType: "text",
+    viewType: "kieKogitoWebviewEditorsDashbuilder",
+    editorEnvelopeLocator: new EditorEnvelopeLocator("vscode", [
+      new EnvelopeMapping({
+        type: "dashbuilder",
+        filePathGlob: "**/*.dash.+(yaml|yml)",
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/src/extension/extension.ts]:
```suggestion
        filePathGlob: "**/*.dash.+(yaml|yml|json)",
```
Can you put `json` here and remove the `EnvelopeMapping` below?

File: packages/vscode-extension-dashbuilder-editor/src/extension/ComponentsHttpServer.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import * as http from "http";
+import * as fs from "fs";
+import * as path from "path";
+
+export class ComponentServer extends LocalHttpServer {
+  private componentsPath: string;
+  private server: http.Server;
+
+  requestListener = (request: any, response: any) => {
+    if (request.url == "/" || request.url == "" || request.url == "index.html") {
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/src/extension/ComponentsHttpServer.ts]:
```suggestion
    if (["/", "", "index.html"].includes(request.url)) {
```
Optional to make it shorter.

File: packages/vscode-extension-dashbuilder-editor/src/extension/ComponentsHttpServer.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        response.writeHead(200);
+        response.end(content, "utf-8");
+      }
+    });
+  };
+
+  constructor(componentsPath: string) {
+    super();
+    this.componentsPath = componentsPath;
+  }
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/src/extension/ComponentsHttpServer.ts]:
```suggestion
  constructor(private readonly componentsPath: string) {
    super();
  }
```
Optional to make it shorter. You can also remove line 24.

**Review by caponetto — COMMENTED** [review]
Hey @jesuino, I'm really glad to see Dashbuilder on VS Code as well, great job!
I've left some comments and minor suggestions inline.
Please check if they make sense.

File: packages/vscode-extension-dashbuilder-editor/package.json [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    "configuration": {
+      "properties": {
+        "kogito.dashbuilder.automaticallyOpenDashboardEditorAlongsideTextEditor": {
+          "default": "Ask next time",
+          "enum": [
+            "Open automatically",
+            "Do not open",
+            "Ask next time"
+          ],
+          "markdownDescription": "When opening dashbuilder files, decide whether or not to open the Dashboard Editor alongside the text editor. Regardless of the configured option, you can always open the dashbuilder Dashboard Editor using the 'Open as Dashboard' button.",
```
Comment (barboras7) [inline packages/vscode-extension-dashbuilder-editor/package.json]:
```suggestion
          "markdownDescription": "When opening Dashbuilder files, decide whether or not to open the Dashboard Editor alongside the text editor. Regardless of the configured option, you can always open the dashbuilder Dashboard Editor using the 'Open as Dashboard' button.",
```

Not sure about this, but for Serverless Workflow files, we use capital letters S and W...WDYT?

**Comment by barboras7** [comment]
Hi @jesuino,

I am not seeing correctly the icon in the VSCode. I guess I should see the icon that you are adding in the PR, but I only see this:
Screenshot from 2022-10-13 08-58-58

Also, when I try to change the setting in the extension settings, it seems that nothing really changes. When I open the file with .dash.yaml extension, the text editor is not opened alongside. How should I apply the setting?
Screenshot from 2022-10-13 16-02-54

File: packages/vscode-extension-dashbuilder-editor/src/extension/ComponentsHttpServer.ts (hunk 9 above)
Comment (jesuino) [reply packages/vscode-extension-dashbuilder-editor/src/extension/ComponentsHttpServer.ts]: TS HACKER!

File: packages/vscode-extension-dashbuilder-editor/src/extension/ComponentsHttpServer.ts (hunk 10 above)
Comment (jesuino) [reply packages/vscode-extension-dashbuilder-editor/src/extension/ComponentsHttpServer.ts]: TS HACKER²

File: packages/vscode-extension-dashbuilder-editor/package.json (hunk 11 above)
Comment (jesuino) [reply packages/vscode-extension-dashbuilder-editor/package.json]: We had the idea of using DashBuilder - but then looks like we abandoned it - I will discuss with the team which one they prefer and start a refactor if the chose to use DashBuilder!

**Comment by jesuino** [comment]
Hello @caponetto Thanks for the suggestions, I applied all the changes, let me know if I am missing something;

Hello @barboras7 I fixed the icon, thanks! I could not reproduce the issue with the editor not opening to the side - but I recall seeing it when I didn't reloaded VSCode after installing the VSIX - could this be the case?

**Comment by caponetto** [comment]
Thanks for the changes, @jesuino!
Can you please add `README.md` ([reference](https://github.com/kiegroup/kie-tools/blob/main/packages/vscode-extension-serverless-workflow-editor/README.md)) and  `CHANGELOG.md` ([reference](https://github.com/kiegroup/kie-tools/blob/main/packages/vscode-extension-yard-editor/CHANGELOG.md)) files?
They will be important when publishing the extension to the stores.

**Comment by barboras7** [comment]
Still not working :disappointed: I tried uninstalling and installing again the extension but no luck. I don't even get the option to choose if I want to open also the text editor. Any idea how to solve this? I thought it might be some issue with some cache but I did not succeed. I'll try to fix it on Monday.

File: packages/vscode-extension-dashbuilder-editor/README.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
+## Kogito Serverless Workflow Editor
```
Comment (caponetto) [inline packages/vscode-extension-dashbuilder-editor/README.md]:
```suggestion
## Dashbuilder Editor
```
😃

Comment (jesuino) [reply packages/vscode-extension-dashbuilder-editor/README.md]: thanks :P

**Comment by jesuino** [comment]
@caponetto thanks, added the files!

@barboras7 thanks for checking it, I am trying to reproduce locally as well!

**Review by caponetto — APPROVED** [review]

File: packages/vscode-extension-kogito-bundle/package.json [hunk 13]
```diff
@@ -46,7 +46,8 @@
   "extensionDependencies": [
     "redhat.vscode-extension-bpmn-editor",
     "redhat.vscode-extension-dmn-editor",
-    "redhat.vscode-extension-serverless-workflow-editor"
+    "redhat.vscode-extension-serverless-workflow-editor",
+    "redhat.vscode-extension-dashbuilder"
```
Comment (caponetto) [inline packages/vscode-extension-kogito-bundle/package.json]:
```suggestion
    "redhat.vscode-extension-dashbuilder-editor"
```
Needs to be the same name defined in the `package.json` file.

**Review by caponetto — APPROVED** [review]

**Comment by jesuino** [comment]
Hello @barboras7  - Guilherme found an issue with the `package.json configuration and it could be related to the issue you saw.

Would you please check the latest version of this PR?

Thanks!

**Comment by barboras7** [comment]
Hi @jesuino,
thank you for trying to fix this. I tried the latest version of the PR and it did not help. Also, I tried cleaning up the VSCode cache but no luck there. 
I think we can merge this PR and if I keep having this issue, I'll try to debug and will create a Jira for it. Thanks!

**Review by barboras7 — APPROVED** [review]

**Review by paulovmr — APPROVED** [review]

**Comment by jesuino** [comment]
thank you, @barboras7 I am sad you faced this issue - I will keep an eye if we see any problem with it we will work on a fix

**Pull request merged**